app/models/namespace.rb

```diff
@@ -99,7 +99,18 @@ def ensure_dir_exist
   end
 
   def rm_dir
-    gitlab_shell.rm_namespace(path)
+    # Move namespace directory into trash.
+    # We will remove it later async
+    new_path = "#{path}+#{id}+deleted"
+
+    if gitlab_shell.mv_namespace(path, new_path)
+      message = "Namespace directory \"#{path}\" moved to \"#{new_path}\""
+      Gitlab::AppLogger.info message
+
+      # Remove namespace directroy async with delay so
+      # GitLab has time to remove all projects first
+      GitlabShellWorker.perform_in(5.minutes, :rm_namespace, new_path)
```

Comment:
@jacobvosmaer I made similar approach for namespace in case when we remove user or group